### PR TITLE
WorkflowManagerActor is finally FSM'd.

### PR DESCRIPTION
**Pull request by mcovarr**
Basic structure should be okay but I haven't finished testing it.

### CODE REVIEW

**Comment by mcovarr** [comment]
The wheel has decided it's @geoffjentry's and @scottfrazer's lucky day

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 1]
```diff
@@ -79,18 +85,19 @@ case class SingleWorkflowRunnerActor(source: WorkflowSourceFiles,
...
+          workflowManager ! WorkflowAbort(id)
+          log.info(s"$tag: Waiting for workflow $id to abort...")
+          Await.ready(donePromise.future, Duration.Inf)
+        }
       }
       stay using data.copy(id = Option(id))
+    case Event(Transition(_, _, WorkflowAborted), data) =>
+      // The .get is safe because the shutdown hook is only registered when the workflow ID
+      // has been set in the state data.
+      log.info(s"$tag: Workflow ${data.id.get} aborted.")
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: that might benefit from an ADTing

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 2]
```diff
@@ -147,32 +147,94 @@ class WorkflowManagerActor(backend: Backend) extends Actor with CromwellActor {
...
+      workflowStore.toMap.values.foreach { _ ! WorkflowActor.AbortWorkflow }
+      goto(Aborting)
+  }
+
+  when (Aborting) {
+    case Event(Transition(workflowActor, _, toState: WorkflowState), _) if toState.isTerminal =>
+      // Remove this terminal actor from the workflowStore and log a progress message.
+      removeTerminalWorkflow(workflowActor)
+      logger.info(s"$tag: Waiting for all workflows to abort (${workflowStore.size} remaining).")
+      // If there are no more workflows to abort we're done, otherwise just stay in the current state.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: The comment is misleading as written, abort messages were already sent to all running workflows.

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 3]
```diff
@@ -79,18 +85,19 @@ case class SingleWorkflowRunnerActor(source: WorkflowSourceFiles,
...
+      stay
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: parens

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Why are you nitpicking yourself? :)

Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Think of these as post its so I don't forget :smile:

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 4]
```diff
@@ -47,12 +47,21 @@ object WorkflowManagerActor {
...
   // restart 500 workflows at exactly the same time crushes the database connection pool.
   lazy val RestartDelay = 200 milliseconds
+
+  sealed trait WorkflowManagerState
+  case object Running extends WorkflowManagerState
+  case object Aborting extends WorkflowManagerState
+  case object Done extends WorkflowManagerState
+
+  sealed trait WorkflowManagerData
+  case object NoData extends WorkflowManagerData
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: The state data should include the workflow store, at which time the write-once class can be deleted.

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: How were you picturing it?

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 3 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Consider it done.

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 5]
```diff
@@ -62,32 +71,23 @@ object WorkflowManagerActor {
...
-          numRemaining = Await.result(result,Duration.Inf).size
-          log.info(s"$tag: Waiting for all workflows to abort ($numRemaining remaining).")
-        }
-        log.info(s"$tag: All workflows aborted.")
-      }
-    })
+  private val donePromise = Promise[Unit]()
+
+  if (abortJobsOnTerminate) {
+    sys.addShutdownHook {
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: why is this here and on the single workflow actor?

Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: These are the shutdown hooks our friends added to support abort on SIGINT.  I did a bit of cleanup related to that in SWRA.  Unfortunately WMA was too crufty to take a similarly minimal change, which motivated the FSM'ing.  This and the CallExecutionActor were the only major actors not to be FSM'd, though @Horneth is already contemplating FSM'ing CAE for reasons unrelated to shutdown hooks.  :smile:

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 1 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: With more state data types and without so many `Option`s :smile:  Haven't thought it through any more than that yet.

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala (hunk 5 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: What I meant was why do you need both hooks? The one in the single actor sends WMA a signal to abort its WF and the one here aborts all workflows. If this is just for run (i forget) wouldn't this be redundant and if this is for both run/server wouldn't the other be redundant?

Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: Ah gotcha.  I hadn't considered that, I think you're right we may be able to get away with only one hook in WMA.  SWRA's `WorkflowAborted` handler would need to be modified to allow for that, but I think that should be done anyway as the current SWRA abort handler looks wrong (it notes that the only workflow it cares about has been aborted and then tells no other actors and hangs out forever doing nothing).

File: src/main/scala/cromwell/util/WriteOnceStore.scala [hunk 6]
```diff
@@ -21,4 +21,21 @@ class WriteOnceStore[K, V] {
...
   def toMap: Map[K, V] = writeOnceStore.readOnlySnapshot().toMap
-}
\ No newline at end of file
+
+  /**
+    * Remove an entry in the store as specified by key.
+    * @param key Key of the entry to remove.
+    * @return `Some(value)` if the entry is present in the store, otherwise `None`.
+    */
+  def remove(key: K): Option[V] = writeOnceStore.remove(key)
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/util/WriteOnceStore.scala]: Something seems strange about making a WriteOnce class which lets you update its values...? Is it more intellectually honest to call this a NoDuplicatesStore?

Come to think of it... isn't this just a mutable Set?

Comment (geoffjentry) [inline src/main/scala/cromwell/util/WriteOnceStore.scala]: I forgot about this (and honestly, I forgot this class was here). You're right, this really bends the spirit of what it is supposed to be, it was supposed to literally be write once per value and that's it

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 7]
```diff
@@ -79,18 +85,19 @@ case class SingleWorkflowRunnerActor(source: WorkflowSourceFiles,
...
-            while(stateName != Done)
-              Thread.sleep(1000)
-            log.info(s"$tag: Workflow $id aborted.")
-          }
-        })
+      if (abortJobsOnTerminate) {
+        sys.addShutdownHook {
+          workflowManager ! WorkflowAbort(id)
+          log.info(s"$tag: Waiting for workflow $id to abort...")
+          Await.ready(donePromise.future, Duration.Inf)
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: This donePromise seems incredibly funky to me. 

It's basically used like a boolean var, but hidden in a Promise to avoid offending FP sensibilities?

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 8]
```diff
@@ -79,18 +85,19 @@ case class SingleWorkflowRunnerActor(source: WorkflowSourceFiles,
...
+        }
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Why not move the `if` block up into the `CromwellActor` (right below the configuration reading) and just override a `customShutdownHook` method here?

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 9]
```diff
@@ -11,8 +11,8 @@ import cromwell.engine.db.DataAccess._
 import cromwell.engine.db.ExecutionDatabaseKey
 import cromwell.engine.db.slick._
 import cromwell.engine.workflow.WorkflowActor.{Restart, Start}
+import cromwell.engine.workflow.WorkflowManagerActor.{WorkflowManagerData, WorkflowManagerState}
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: I guess because the Object methods are used in the Class? 

Has anyone ever opined on whether its better to have these all the way up here, or as a line in the class definition?

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 10]
```diff
@@ -62,32 +71,23 @@ object WorkflowManagerActor {
  * WorkflowOutputs: Returns a `Future[Option[binding.WorkflowOutputs]]` aka `Future[Option[Map[String, WdlValue]]`
  *
  */
-class WorkflowManagerActor(backend: Backend) extends Actor with CromwellActor {
+class WorkflowManagerActor(backend: Backend) extends LoggingFSM[WorkflowManagerState, WorkflowManagerData] with CromwellActor {
   import WorkflowManagerActor._
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: Oh wait, you ARE importing it here anyway!

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala (hunk 9 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: I personally strongly prefer having just about everything up at the top unless there's a good reason not to (e.g. pulling a specific implicit into scope)

File: src/main/scala/cromwell/util/WriteOnceStore.scala (hunk 6 above)
Comment (cjllanwarne) [inline src/main/scala/cromwell/util/WriteOnceStore.scala]: I actually just saw a TODO higher up this PR saying "... and this lets us remove the WriteOnceStore!" so hopefully this is about to disappear?

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala (hunk 4 above)
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: woohoo!

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 7 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: I think it is structured this way because the Promise will be fulfilled asynchronously and there's no way to say "wait for a boolean var to become `true`"?

Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: I guess there's a reason why `Await.ready(..., Duration.Inf)` is better than `while(!boolVar) sleep(1)` - I'd be happy enough to hear "behind-the-scenes-FP-magic, don't worry about it"

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: If I understand correctly, `sleep(1)` doesn't actually yield control of the thread in Akka-land, whereas the `Await.ready()` would.  I do agree with you that it is funky looking, though.

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: @cjllanwarne I'm probably missing a variety of other bits but one thing is that `Await.ready` calls `blocking {}` under the hood, which gives a heads up to your execution context that danger is incoming so it can try to manage things a bit better. You also might get better granularity than in your polling loop.

Another angle which seems trivial but I'd argue it's very much not is that it makes it more clear that blocking is happening. `Await` pops out like a giant red alarm bell, at least to me.

Iff I haven't missed anything key the differences would be relatively minor.

@scottfrazer I believe both `Awaits` really do block the thread, but they should both have that `blocking` keyword in there (nothing prevents someone from doing that themselves) which helps make things more sane. That's part of why they're also viewed with suspicion.

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Hmm I thought something was up... but it's very unlike @mcovarr to write blocking code so I thought I wasn't really understanding something

Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: As far as I understand the contract (and consistent with what I've observed), the JVM tries to execute all shutdown hooks and then exit.  So this actually needs to block, otherwise the JVM would shut down as soon as the code here finished without waiting for workflow(s) to abort.  I'll add a comment to this effect.

File: src/main/scala/cromwell/util/WriteOnceStore.scala (hunk 6 above)
Comment (mcovarr) [inline src/main/scala/cromwell/util/WriteOnceStore.scala]: yup

File: src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 7 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Ah that makes a lot more sense

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala (hunk 9 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: These two are actually used as type variables in the *definition* of the class.  But I can `_` this and get rid of the other import below.

**Comment by mcovarr** [comment]
@dgtester just a heads up that I did some refactoring in the area of your recent PR that might at first look like I deleted your work, but actually that logic has been consolidated to `WorkflowManagerActor`.  I've tested that SIGINT still works in both "run" and "server" use cases, but please let me know if you have any questions or concerns.

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 11]
```diff
@@ -47,12 +48,33 @@ object WorkflowManagerActor {
...
   // How long to delay between restarting each workflow that needs to be restarted.  Attempting to
   // restart 500 workflows at exactly the same time crushes the database connection pool.
   lazy val RestartDelay = 200 milliseconds
+
+  sealed trait WorkflowManagerState
+  case object Running extends WorkflowManagerState
+  case object Aborting extends WorkflowManagerState
+  case object Done extends WorkflowManagerState
+
+  type WorkflowActorRef = ActorRef
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: What is this buying you?

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 12]
```diff
@@ -320,28 +392,38 @@ class WorkflowManagerActor(backend: Backend) extends Actor with CromwellActor {
...
+    */
+  private def submitWorkflow(source: WorkflowSourceFiles, replyTo: Option[ActorRef], id: Option[WorkflowId]): WorkflowManagerData = {
     val workflowId: WorkflowId = id.getOrElse(WorkflowId.randomId())
-    log.info(s"$tag submitWorkflow input id = $id, effective id = $workflowId")
+    logger.info(s"$tag submitWorkflow input id = $id, effective id = $workflowId")
     val isRestart = id.isDefined
 
-    val workflowActor = for {
-      descriptor <- Future.fromTry(Try(WorkflowDescriptor(workflowId, source)))
+    case class ActorAndStateData(actor: WorkflowActorRef, data: WorkflowManagerData)
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: w00t class and not tuple!

**Comment by geoffjentry** [comment]
:+1:

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala (hunk 11 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: Preexisting, I just moved it up a ways.  Arguably more beauteous.

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: I don't really care BTW, just so long as you think it's providing value

Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: I don't think it adds a ton of value, but it was already here so (shrug)

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: It's not hard to remove :)

**Comment by scottfrazer** [comment]
:+1:
